src/Guzzle/CiviCrmClient.php

```diff
@@ -33,51 +42,165 @@ public function __construct(ClientInterface $client, string $apiKey, string $sit
         $this->siteKey = $siteKey;
     }
 
-    public function subscribe(string $email, array $preferences, string $firstName = null, string $lastName = null) : PromiseInterface
+    private function storePreferencesUrl(int $contactId, string $preferencesUrl) : PromiseInterface
     {
         return $this->client->sendAsync($this->prepareRequest('POST'), $this->options([
             'query' => [
                 'entity' => 'Contact',
                 'action' => 'create',
-                'json' => array_filter([
+                'json' => [
+                    'contact_id' => $contactId,
+                    self::FIELD_PREFERENCES_URL => $preferencesUrl,
+                ],
+            ],
+        ]))->then(function (Response $response) {
+            return $this->prepareResponse($response);
+        })->then(function ($data) {
+            return ['contact_id' => $data['id']];
+        });
+    }
+
+    public function subscribe(string $identifier, array $preferences, string $preferencesUrl, string $firstName = null, string $lastName = null, array $preferencesBefore = null) : PromiseInterface
+    {
+        return $this->client->sendAsync($this->prepareRequest('POST'), $this->options([
+            'query' => [
+                'entity' => 'Contact',
+                'action' => 'create',
+                'json' => [
                     'contact_type' => 'Individual',
-                    'email' => $email,
-                    'first_name' => $firstName,
-                    'last_name' => $lastName,
-                ]),
+                    !is_null($preferencesBefore) ? 'contact_id' : 'email' => $identifier,
+                    'first_name' => $firstName ?? '',
+                    'last_name' => $lastName ?? '',
+                    self::FIELD_PREFERENCES_URL => $preferencesUrl,
+                    // Interpret submission as confirmation of desire to receive bulk emails.
+                    'is_opt_out' => 0,
```

Comment:
If subscribing then we need to ensure `is_opt_out` is set to `0`.

Reply:
I need to consider the case that someone attempts to signup for our newsletter that has previously opted out of receiving bulk email. If a contact in civi opts out of bulk email then they are not removed from any of their groups but civi blocks them from being recipients for any of our mailings.

What I am doing here is to assume that if a user is going through the sign up process for the etoc we can lift the opt out status for that user.

If we are happy with that then this work is ready to go.